src/rust/cryptography-x509-validation/src/backend.rs

```diff
@@ -0,0 +1,20 @@
+// This file is dual licensed under the terms of the Apache License, Version
+// 2.0, and the BSD License. See the LICENSE file in the root of this repository
+// for complete details.
+
+//! Behavioral typing for a "backend" that provides cryptographic operations.
+
+use cryptography_x509::certificate::Certificate;
+
+pub trait Backend {
+    /// A public key type for this backend.
+    type Key;
+
+    /// Extracts the public key from the given `Certificate` in
+    /// a `Key` format known by the backend.
+    fn public_key(cert: &Certificate) -> Self::Key;
+
+    /// Verifies the signature on `Certificate` using the given
+    /// `Key`.
+    fn is_signed_by(cert: &Certificate, key: Self::Key) -> bool;
```

Comment:
These probably should take `&self`

Reply:
I think they should take `&self` both in the trait and also in the impl. What's the coherence issue?

Reply:
Both in the trait and impl, yep, but the impl should be _on_ something that isn't `Certificate`, right? The coherence issue is that `Certificate` is a third-party type (since we're doing the validation in its own crate), so there can only be one trait impl for it.

In other words, is this what you're thinking?

```rust
trait CryptoOps {
  fn public_key(&self, cert: &Certificate) -> Self::Key;
}

struct FooCrypto;
impl CryptoOpts for FooCrypto {
  fn public_key(&self, cert: &Certificate) -> Self::Key {
    todo!()
  }
}
```

That makes sense to me, just wanted to make sure I understood correctly.

Reply:
Yes, that's right. We want to allow for stateful CryptoOps.

Reply:
Done in f16141ebc -- `CryptoOps` can now access its associated state, at the cost of slightly uglier policy/profile APIs 🙂